### PR TITLE
Feature/ubs

### PR DESCRIPTION
**Pull request by JimHolman1199**
related to: 
https://github.com/ita-social-projects/GreenCity/issues/2200
https://github.com/ita-social-projects/GreenCity/issues/2197
https://github.com/ita-social-projects/GreenCity/issues/2198
https://github.com/ita-social-projects/GreenCity/issues/2199
https://github.com/ita-social-projects/GreenCity/issues/2196

### CODE REVIEW

**Review by KirillAbramo — APPROVED** [review]

File: src/app/component/ubs/components/ubs-input-error/ubs-input-error.component.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  @Input() public streetFieldValue: string;
+  @Input() public houseNumberFieldValue: string;
+
+  public errorMessage = '';
+  private getErrorMsg = {
+    required: () => 'Це поле є обов\'язкове до заповнення',
+    email: () => this.emailFieldValue ? 'Введіть коректний email' : 'Це поле  обов\'язкове до заповнення',
+    minlength: () => this.phoneNumberFieldValue ? 'Введіть номер телефону повністю' : 'Мінімальна кількість символів: 3',
+    maxlength: () => 'Перевищена максимальна кількість символів',
+    pattern: () => 'Поле містить заборонені символи'
```
Comment (mr-pavlenko) [inline src/app/component/ubs/components/ubs-input-error/ubs-input-error.component.ts:25]: Do we need translation for this field?

File: src/app/component/ubs/components/ubs-input-error/ubs-input-error.component.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  @Input() public formElement: FormControl;
+  @Input() public certificateFieldValue: string;
+  @Input() public firstNameFieldValue: string;
+  @Input() public lastNameFieldValue: string;
+  @Input() public emailFieldValue: string;
+  @Input() public phoneNumberFieldValue: string;
+  @Input() public streetFieldValue: string;
+  @Input() public houseNumberFieldValue: string;
+
+  public errorMessage = '';
```
Comment (mr-pavlenko) [inline src/app/component/ubs/components/ubs-input-error/ubs-input-error.component.ts:19]: Why do we need empty string for error?

File: src/app/component/ubs/components/ubs-input-error/ubs-input-error.component.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  constructor() {}
+
+  ngOnChanges() {
+    this.getType();
+  }
+
+  private getType() {
+    Object.keys(this.formElement.errors).forEach(error => {
+      return this.errorMessage = this.getErrorMsg[error]();
```
Comment (mr-pavlenko) [inline src/app/component/ubs/components/ubs-input-error/ubs-input-error.component.ts:36]: Can we have multiple errors? If so we always re-write msg.  If no we don't need foreach here.

File: src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.html [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+      <div class="form-group comment">
+        <h3>Коментар</h3>
+        <textarea formControlName="orderComment" class="shadow-none form-control" rows="5"
+          placeholder="Будь-які деталі які Ви б хотіли повідомити додатково щодо цього замовлення"></textarea>
+      </div>
+      <div class="w-100 d-flex justify-content-end">
+        <div>
+        <button class="secondary-global-button btn" matStepperPrevious>Відмінити</button>
+        <button type="submit" class="primary-global-button btn" matStepperNext [disabled]="onSubmit">Далі</button>
+      </div>
```
Comment (mr-pavlenko) [inline src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.html:185]: All text should be translated or moved to a const file.

File: src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.scss [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+  width: 25%;
+
+  p {
+    display: flex;
+    justify-content: space-between;
+    color: var(--ubs-secondary-grey);
+    padding-right: 0.5rem;
+  }
+
+  span {
```
Comment (mr-pavlenko) [inline src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.scss:124]: Why do you need span as a block element?

File: src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.scss [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+
+  .custom-control {
+    padding-left: 1.4rem;
+  }
+
+  label {
+    margin: 0;
+  }
+
+  input {
```
Comment (mr-pavlenko) [inline src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.scss:151]: use classes

File: src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.scss [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+    margin: 0 24px 0 0;
+    border: 1px solid var(--ubs-primary-green);
+    background: none;
+    color: var(--ubs-primary-green);
+  }
+
+  button:hover {
+    background: var(--ubs-primary-green);
+  }
+}
```
Comment (mr-pavlenko) [inline src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.scss:297]: Please use nesting for SCCS and extends if you have the same styles.

File: src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.ts [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+import { UserOrder } from '../../models/ubs.model';
+import { IOrder, IUserOrder } from '../../models/ubs.interface';
+
+@Component({
+  selector: 'app-ubs-order-details',
+  templateUrl: './ubs-order-details.component.html',
+  styleUrls: ['./ubs-order-details.component.scss'],
+})
+export class UBSOrderDetailsComponent implements OnInit, OnDestroy {
+  showTotal = 0;
```
Comment (mr-pavlenko) [inline src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.ts:17]: Is better to use an object to split the logic of variables.

File: src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.ts [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+      bagPriceClothesXL: `${this.orders.allBags[1].price} грн`,
+      bagServiceClothesM: this.orders.allBags[2].name,
+      bagSizeClothesM: `${this.orders.allBags[2].capacity} л`,
+      bagPriceClothesM: `${this.orders.allBags[2].price} грн`,
+    });
+    this.points = this.orders.points;
+  }
+
+  calculateTotal(): void {
+    this.total =
```
Comment (mr-pavlenko) [inline src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.ts:122]: Is this function private?

File: src/app/component/ubs/components/ubs-order-form/ubs-order-form.component.scss [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  .mat-step-label.mat-step-label-active {
+    color: #64727d;
+  }
+
+  .mat-step-label.mat-step-label-selected {
+    color: var(--primary-green);
+  }
+
+  .mat-step-icon {
+    background-color: #e3e6e8;
```
Comment (mr-pavlenko) [inline src/app/component/ubs/components/ubs-order-form/ubs-order-form.component.scss:17]: use variable

File: src/app/component/ubs/components/ubs-order-form/ubs-order-form.component.scss [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    font-size: 16px;
+    font-weight: bold;
+  }
+
+  .mat-step-icon-state-number.mat-step-icon-selected {
+    color: #fff;
+  }
+
+  .mat-icon {
+    color: #64727d;
```
Comment (mr-pavlenko) [inline src/app/component/ubs/components/ubs-order-form/ubs-order-form.component.scss:35]: for all colors use variables.

File: src/app/component/ubs/components/ubs-personal-information/ubs-personal-information.component.ts [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        takeUntil(this.destroy)
+      )
+      .subscribe((data: PersonalData[]) => {
+        this.personalData = data[data.length - 1];
+        this.initForm();
+      });
+  }
+
+  ngOnDestroy(): void {
+    this.shareFormService.objectSource.unsubscribe();
```
Comment (mr-pavlenko) [inline src/app/component/ubs/components/ubs-personal-information/ubs-personal-information.component.ts:117]: Resolve your destroy.

**Review by mr-pavlenko — CHANGES_REQUESTED** [review]

File: src/app/component/ubs/components/ubs-input-error/ubs-input-error.component.ts (hunk 2 above)
Comment (JimHolman1199) [reply src/app/component/ubs/components/ubs-input-error/ubs-input-error.component.ts]: fixed

File: src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.scss (hunk 5 above)
Comment (JimHolman1199) [reply src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.scss]: fixed

File: src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.ts (hunk 9 above)
Comment (JimHolman1199) [reply src/app/component/ubs/components/ubs-order-details/ubs-order-details.component.ts:122]: fixed

File: src/app/component/ubs/components/ubs-order-form/ubs-order-form.component.scss (hunk 11 above)
Comment (JimHolman1199) [reply src/app/component/ubs/components/ubs-order-form/ubs-order-form.component.scss]: fixed

File: src/app/component/ubs/components/ubs-personal-information/ubs-personal-information.component.ts (hunk 12 above)
Comment (JimHolman1199) [reply src/app/component/ubs/components/ubs-personal-information/ubs-personal-information.component.ts:117]: fixed

File: src/app/component/ubs/components/ubs-order-form/ubs-order-form.component.scss (hunk 10 above)
Comment (JimHolman1199) [reply src/app/component/ubs/components/ubs-order-form/ubs-order-form.component.scss]: fixed

**Pull request merged**